[TextureMapper] Add precision hint to samplerExternalOES

zhani opened the pull request:
#### 2db0856f7d83622886b58f16c19a6cc748cc1b63
<pre>
[TextureMapper] Add precision hint to samplerExternalOES
<a href="https://bugs.webkit.org/show_bug.cgi?id=279586">https://bugs.webkit.org/show_bug.cgi?id=279586</a>

Reviewed by Miguel Gomez.

Some Android emulator versions fail to render because glShaderSource in
TextureMapperShaderProgram fails with the error 0x505 GL_OUT_OF_MEMORY.
It turns out that the Android emulator&apos;s OpenGL ES implementation does
not implicitly define any precision for samplerExternalOES.

Adding a precision hint to samplerExternalOES across all platforms is
harmless since OpenGL ES implementations are not required to use hint
information. This change is necessary for WPEView/Android.

* Source/WebCore/platform/graphics/texmap/TextureMapperShaderProgram.cpp:

Canonical link: <a href="https://commits.webkit.org/283627@main">https://commits.webkit.org/283627@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a43bd562d004aaeab643624909179cf922d27b36

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66648 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46023 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19270 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70682 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/17781 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68766 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53822 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17545 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53402 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11994 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69715 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42371 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57675 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34071 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39042 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15066 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16135 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60957 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15407 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72384 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10605 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14767 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60731 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10637 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57739 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61064 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14769 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8729 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2348 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41830 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42907 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44090 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42650 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

magomez commented on Source/WebCore/platform/graphics/texmap/TextureMapperShaderProgram.cpp:
```diff
@@ -151,6 +151,7 @@ static const char* vertexTemplateCommon =
     GLSL_DIRECTIVE(ifdef ENABLE_TextureExternalOES) \
         GLSL_DIRECTIVE(extension GL_OES_EGL_image_external : require) \
         GLSL_DIRECTIVE(define SamplerExternalOESType samplerExternalOES) \
+        STRINGIFY(precision mediump samplerExternalOES;\n) \
```
is there any reason why you're not using GLSL_DIRECTIVE as the rest of the code?

zhani replied:
Yes, GLSL_DIRECTIVE  adds # to the beginning of the line.
With GLSL_DIRECTIVE line would become
`#precision mediump samplerExternalOES;`
and that's not valid shader code

magomez replied:
ah, you're right! But please, follow the same format as the rest of the file:
STRINGIFY(
    precision mediump samplerExternalOES;
)

zhani replied:
It's inside a macro so it would be a slightly different
```
STRINGIFY( \
    precision mediump samplerExternalOES;\n \
) \
```
I think one-liner is more clean looking, WDYT?

magomez replied:
The options don't look that different, but I think we should try to use the same structure in all the code.

zhani replied:
ok, updated formatting accordingly

zhani committed:
[TextureMapper] Add precision hint to samplerExternalOES
https://bugs.webkit.org/show_bug.cgi?id=279586

Reviewed by Miguel Gomez.

Some Android emulator versions fail to render because glShaderSource in
TextureMapperShaderProgram fails with the error 0x505 GL_OUT_OF_MEMORY.
It turns out that the Android emulator's OpenGL ES implementation does
not implicitly define any precision for samplerExternalOES.

Adding a precision hint to samplerExternalOES across all platforms is
harmless since OpenGL ES implementations are not required to use hint
information. This change is necessary for WPEView/Android.

* Source/WebCore/platform/graphics/texmap/TextureMapperShaderProgram.cpp:

Canonical link: https://commits.webkit.org/283627@main
